include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1145 @@
+     */
+    const uint64_t multiset_conf;
+    /**
+     * Getter for the number of a given charge state in the multiset configuration.
+     *
+     * @tparam cs Charge state to count the number of occurrences in the projector state of.
+     * @return The number of occurrences of the given charge state in the multiset charge configuration. For a neutral
+     * charge, the number of occurrences is inferred by considering the size of the cluster in the projector state.
+     */
+    template <sidb_charge_state cs>
+    [[nodiscard]] constexpr inline uint64_t get_count() const noexcept
+    {
+        switch (cs)
+        {
+            case sidb_charge_state::NEGATIVE: return multiset_conf >> 32ull;
+            case sidb_charge_state::POSITIVE: return multiset_conf & 0xFFFFFFFF;
+            default: return get_cluster_size(cluster) - (multiset_conf >> 32ull) - (multiset_conf & 0xFFFFFFFF);
+        }
+    }
+};
+/**
+ * Projector state pointers are unique.
+ */
+using sidb_cluster_projector_state_ptr = std::unique_ptr<sidb_cluster_projector_state>;
+/**
+ * The electrostatic potential bounds required for the *Ground State Space* algorithm. As the domain in
+ * which our potential bounds live are simply the real numbers, we may think of the lower bound and upper bound domains
+ * to be separate partial order structures on the real number line, inverse to each other. The *Ground State Space*
+ * algorithm requires the properties of a lower semi-lattice for these domains, ie. all finite meets must exist. This is
+ * implemented for the lower and upper bound respectively simply by taking a minimum or a maximum. One may think of
+ * meets as follows, which is very relevant to intention of their application: a meet, or greatest lower bound, is the
+ * *maximal information* common to a set (of potential bounds). This semantic operation is essential to the *Ground
+ * State Space* algorithm, which thus envelops without loss of accuracy.
+ */
+enum class bound_direction : uint8_t
+{
+    LOWER = 0,
+    UPPER
+};
+/**
+ * The respective *top* elements of the lower semi-lattices in which our potential bounds live, ie., the respective
+ * elements of *most* information. This means that any meet with a top element (weakly) reduces the information, and
+ * thus any meet computed in iteration through binary application may start out with this element.
+ *
+ * @tparam bound Potential bound domain to return the element of most information of.
+ * @return The element of most information respective to the potential bound domain.
+ */
+template <bound_direction bound>
+static constexpr inline double potential_bound_top() noexcept
+{
+    if constexpr (bound == bound_direction::LOWER)
+    {
+        return std::numeric_limits<double>::infinity();
+    }
+    else if constexpr (bound == bound_direction::UPPER)
+    {
+        return -std::numeric_limits<double>::infinity();
+    }
+}
```

Comment:
Same as before with the missing return statement. Either add ``else`` or write it like this:
```suggestion
template <bound_direction bound>
static constexpr inline double potential_bound_top() noexcept
{
    if constexpr (bound == bound_direction::LOWER)
    {
        return std::numeric_limits<double>::infinity();
    }
    return -std::numeric_limits<double>::infinity();
}
```